### ISSUE TITLE
Scanning for projects on Windows takes forever

### ISSUE THREAD

**Issue by CraftSpider** [dnlmlr/cargo-clean-all#9]
I used this tool on Linux and liked it, so was also trying to use it on my Windows machine. However, when I ran it on a (relatively large) git projects directory, it takes a very long time - not sure exactly how long yet, but at least 5 minutes and still going.

Running in verbose mode doesn't give me any hint as to what's taking so long - I definitely have some projects with large subdirectory trees, but I'd still not expect it to take near this long. I can do some local debugging if you want any more information.

**Comment by dnlmlr** [comment]
Hi! Could you give some more information on the context concerning the following points:
- Is the projects directory you want to clean located on HDD or SSD storage?
- What is the active disk utilization shown in the Task-Manager? (The active time value for the drive in question on the performance tab)
- How many files and subdirectories are in your projects directory? (This should be shown in the properties of the project directory)

As a reference, scanning my projects directory on fast SSD NVME storage with about 193k files and 26748 directories takes about 3 seconds. I tested this with Windows 11 and a Ryzen 7900x CPU.

If the disk utilization on your machine is at or near 100%, it might just be reaching the limits of the scanning technique that is used here. This tool is using the default std directory listing functions with optionally (default enabled) multithreading. Especially the multithreading could potentially even harm the performance when you are using a HDD. You could try to set the number of threads manually to a lower number using `-t`

**Comment by CraftSpider** [comment]
I'm using an HDD, disk utilization is at 100%, and the total number of files is ~500k, directories ~75k, though I *should* be excluding ~376k and ~53k of those, respectively, unless the std method still scans those excluded directories?

**Comment by dnlmlr** [comment]
Thanks for the answer. The full scan on HDD storage for this many directories will definitely take a pretty long time, especially if the cache is cold. To do that kind of a scan in a more reasonable time, the program would have to do something like what [WizTree](https://diskanalyzer.com/) or [Everything](https://www.voidtools.com/) is doing. I would love to look into that at some point because it is pretty interesting in general but that is not really in scope for this project right now.

More relevant to your situation, you stated that *it should be excluding* a major part of that, which indeed should make it significantly faster. I assume you might have used something like `cargo clean-all --ignore myprojects/cpp --ignore myprojects/javascript myprojects` in order to clean the `myprojects` directory but skip over certain subdirectories and their respective fs trees. 
If this is the case, that did not do what you were expecting. The help entry for `--ignore` is currently:
```
--ignore <IGNORE>    Directories that should be ignored by default, including subdirectories. 
This will still detect the projects in those directories, but mark them to not be cleaned
```
The important part here being that projects will still be scanned, just not marked for cleaning.
I will admit that the argument is confusingly named and the possibility to actually skip over subdirectories completely would be extremely helpful here. The `--ignore` argument was designed together with the interactive mode in order to provide the same way of not cleaning certain projects while not using interactive mode. And since I personally have my stuff on pretty fast SSD storage and my project dir is a lot smaller, I never ran into the situation where fully skipping subdirs would make a difference.

**Actions**
I'll add a `--skip` option that will take arguments like `--ignore` but cause the scanner to fully skip over the specified directories, speeding up the scan time in the progress. While adding that I'll try to make the argument descriptions in the help page be clear on what `--ignore` and `--skip` do exactly and how it is different. I still want to keep the `--ignore` option to show how much potentially freeable space is being ignored

**Comment by CraftSpider** [comment]
Awesome, thank you! I even read and knew the interactive mode might include them, I just didn't think hard enough about what that meant in terms of performance, so I'll take some blame for that. A `--skip` command sounds just like what I want, thank you! Until then I'll just manually handle it, or do some workaround moving the directory or something.

**Comment by dnlmlr** [comment]
Hey! The `--skip` command is implemented in the new version `0.6.0`, simply doing a `cargo install cargo-clean-all` should do the trick for updating.
Let me know if this helps with the scanning time.

**Comment by CraftSpider** [comment]
Tried `--skip` and it worked amazing - tried with a similar number of directories, but skipping the largest ones, and it finished in less than 10 seconds. Thank you for the quick response, and the useful tool!
